File: java/atcoder/caddi2018b_c.c
#include <stdio.h>

int main(void) {
    long n;
    long p;
    scanf("%ld %ld", &n, &p);
    long remP = p;

    if (n == 1) {
        printf("%ld\n", p);
        return 0;
    }

    long answer = 1;

    long cnt2 = 0;
    while (remP % 2 == 0) {
        remP /= 2;
        cnt2++;
    }
    for (long j = 0; j < cnt2 / n; j++) {
        answer *= 2;
    }

    for (long i = 3; i <= remP && remP > 1; i += 2) { // prime
        long cnt = 0;
        while (remP % i == 0) {
            remP /= i;
            cnt++;
        }
        for (long j = 0; j < cnt / n; j++) {
            answer *= i;
        }
    }
    printf("%ld\n", answer);
    return 0;
}

File: java/atcoder/ab/caddi2018b_c.java
import java.math.*;
import java.util.*;

class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String[] np = scanner.nextLine().split(" ", 2);
        long n = Long.parseLong(np[0]);
        long p = Long.parseLong(np[1]);
        long remP = p;
        long limit = (long)Math.sqrt(p) + 1;

        if (n == 1) {
            System.out.println(p);
            return;
        }

        long answer = 1;

        for (long i = 2; i <= limit; i++) { // prime
            long cnt = 0;
            while (remP % i == 0) {
                remP /= i;
                cnt++;
            }
            for (long j = 0; j < cnt / n; j++) {
                answer *= i;
            }
            if (remP == 1) {
                break;
            }
        }
        System.out.println(answer);
    }
}
